### PR TITLE
🔀  Reimplement Messages and Apps

### PR DESCRIPTION
**Pull request by RomaricMourgues**

### CODE REVIEW

File: twake/backend/node/src/core/platform/services/phpnode/index.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
+import { FastifyPluginCallback, FastifyReply, FastifyRequest } from "fastify";
+import { FastifyInstance } from "fastify/types/instance";
+import { IncomingMessage, Server, ServerResponse } from "http";
+import { Consumes, TwakeService } from "../../framework";
+import WebServerAPI from "../webserver/provider";
+import PhpNodeAPI from "./provider";
+
+@Consumes(["webserver"])
+export default class WebServerService extends TwakeService<PhpNodeAPI> implements PhpNodeAPI {
```
Comment (chamerling) [inline twake/backend/node/src/core/platform/services/phpnode/index.ts:9]: Rename `WebServerService`  to `PhpNode`

File: twake/backend/node/src/core/platform/services/websocket/index.ts [hunk 2]
```diff
@@ -33,21 +35,20 @@ export default class WebSocket extends TwakeService<WebSocketAPI> {
...
-      if (secret && token === secret) {
-        const body = request.body as { room: string; data: any };
-        const room = body.room;
-        const data = body.data;
-        this.service.getIo().to(room).emit("realtime:event", { name: room, data: data });
-        reply.send({});
-      } else {
-        reply.code(401).send({});
-      }
+    phpnode.register((internalServer, _, next) => {
```
Comment (chamerling) [inline twake/backend/node/src/core/platform/services/websocket/index.ts:38]: Are you sure that you need to do it this side?
Why not having this code in the phpnode service? The phpnode requires websocket, and then you can expose the REST API from it, and consume the websocket service to publish event to the room

File: twake/backend/node/src/services/channels/index.ts [hunk 3]
```diff
@@ -28,6 +33,19 @@ export default class ChannelService extends TwakeService<ChannelServiceAPI> {
       next();
     });
 
+    phpnode.register((internalServer, _, next) => {
+      internalServer.route({
+        method: "GET",
+        url: "/companies/:company_id/workspaces/:workspace_id/channels/:id/members/:member_id",
```
Comment (chamerling) [inline twake/backend/node/src/services/channels/index.ts:39]: This route already exists

File: twake/backend/node/src/services/channels/index.ts [hunk 4]
```diff
@@ -28,6 +33,19 @@ export default class ChannelService extends TwakeService<ChannelServiceAPI> {
       next();
     });
 
+    phpnode.register((internalServer, _, next) => {
```
Comment (chamerling) [inline twake/backend/node/src/services/channels/index.ts:36]: Same here, this code must go in the phpnode service

File: twake/backend/node/src/services/channels/services/tab/service.ts [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+      this.service.init && (await this.service.init());
+    } catch (err) {
+      console.error("Can not initialize channel tab service");
+    }
+
+    return this;
+  }
+
+  async save(tab: Tab, context: ChannelExecutionContext): Promise<SaveResult<Tab>> {
+    return new SaveResult("none", null, OperationType.CREATE);
```
Comment (chamerling) [inline twake/backend/node/src/services/channels/services/tab/service.ts:34]: the string value should be `"tab"` here and in all the results below

File: twake/backend/node/src/services/channels/services/tab/service.ts [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+  list(
+    pagination: Pagination,
+    options: ListOptions,
+    context: ChannelExecutionContext,
+  ): Promise<ListResult<Tab>> {
+    return this.service.list(pagination, options, context);
+  }
+
+  async listUserChannels(
```
Comment (chamerling) [inline twake/backend/node/src/services/channels/services/tab/service.ts:54]: You can remove it

File: twake/backend/node/src/services/channels/services/tab/service.ts [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    return new ListResult("none", []);
+  }
+
+  onUpdated(
+    channel: Channel,
+    tab: Tab,
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    updateResult: UpdateResult<Tab>,
+  ): void {
+    console.log("Member updated", tab);
```
Comment (chamerling) [inline twake/backend/node/src/services/channels/services/tab/service.ts:68]: "Tab created" (even if not used for now)

File: twake/backend/node/src/services/channels/web/controllers/member.ts [hunk 8]
```diff
@@ -168,6 +168,27 @@ export class ChannelMemberCrudController
...
+  /**
+   *
+   * @param request Private exists
+   * @param reply
+   */
+  async exists(
+    request: FastifyRequest<{ Params: ChannelMemberParameters }>,
+    reply: FastifyReply,
+  ): Promise<Response> {
+    const resource = await this.service.get(
```
Comment (chamerling) [inline twake/backend/node/src/services/channels/web/controllers/member.ts:181]: I think I have it in my current PR

File: twake/backend/node/src/services/channels/index.ts (hunk 3 above)
Comment (RomaricMourgues) [reply twake/backend/node/src/services/channels/index.ts:39]: No here this is the internal route (without /service/channels/v1/ prefix)
The final route is "/private/companies/etc"

I need the phpnode service to provide a new set of routes prefixed /private for internal communication (and so not open to the world)

File: twake/backend/node/src/core/platform/services/websocket/index.ts (hunk 2 above)
Comment (RomaricMourgues) [reply twake/backend/node/src/core/platform/services/websocket/index.ts]: Done

File: twake/backend/node/src/services/channels/index.ts [hunk 9]
```diff
@@ -28,6 +33,18 @@ export default class ChannelService extends TwakeService<ChannelServiceAPI> {
       next();
     });
 
+    /**
+     * Register private calls from php
+     */
+    phpnode.register({
```
Comment (chamerling) [inline twake/backend/node/src/services/channels/index.ts:39]: You still need to avoid to call phpnode from here but phpnode can call channels.

**Pull request merged**